### PR TITLE
Warn on missing custom texture folder

### PR DESCRIPTION
**Pull request by Zexaron**
Except I don't know yet of a better way make it stop proceeding, currently it will just go ahead and try loading textures from the nonexisting path, it won't find any so it will report that it loaded 0 textures in 0 seconds.

There could be some check to see if the DoFileSearch found anything. For example checking `if filenames == 0` - but what would be the command to stop it?

I was looking into `HiresTextures::Shutdown()` but I thought it wasn't going to do what I think it would do.

### CODE REVIEW

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 1]
```diff
@@ -45,9 +45,7 @@ static std::thread s_prefetcher;
 
 static const std::string s_format_prefix = "tex1_";
 
-HiresTexture::Level::Level() : data(nullptr, SOIL_free_image_data)
-{
-}
+HiresTexture::Level::Level() : data(nullptr, SOIL_free_image_data) {}
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: Unnecessary change.

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 2]
```diff
@@ -463,16 +461,18 @@ std::unique_ptr<HiresTexture> HiresTexture::Load(const std::string& base_filenam
   const Level& first_mip = ret->m_levels[0];
   if (first_mip.width * height != first_mip.height * width)
   {
-    ERROR_LOG(VIDEO, "Invalid custom texture size %ux%u for texture %s. The aspect differs "
-                     "from the native size %ux%u.",
+    ERROR_LOG(VIDEO,
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: Unnecessary change (and lint will probably complain on this).

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 3]
```diff
@@ -463,16 +461,18 @@ std::unique_ptr<HiresTexture> HiresTexture::Load(const std::string& base_filenam
...
+              "from the native size %ux%u.",
               first_mip.width, first_mip.height, first_mip_filename.c_str(), width, height);
   }
 
   // Same deal if the custom texture isn't a multiple of the native size.
   if (width != 0 && height != 0 && (first_mip.width % width || first_mip.height % height))
   {
-    ERROR_LOG(VIDEO, "Invalid custom texture size %ux%u for texture %s. Please use an integer "
-                     "upscaling factor based on the native size %ux%u.",
+    ERROR_LOG(VIDEO,
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: Again, this change is unnecessary.

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 4]
```diff
@@ -543,19 +543,36 @@ bool HiresTexture::LoadTexture(Level& level, const std::vector<u8>& buffer)
 
 std::string HiresTexture::GetTextureDirectory(const std::string& game_id)
 {
-  const std::string texture_directory = File::GetUserPath(D_HIRESTEXTURES_IDX) + game_id;
+  const std::string cstex_altid = game_id.substr(0, 3);  // 3-character region-free version
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: What does `cstex_altid` mean? The variable naming is way, way too obscure...

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 5]
```diff
@@ -543,19 +543,36 @@ bool HiresTexture::LoadTexture(Level& level, const std::vector<u8>& buffer)
...
+  const std::string cstex_rootdir = File::GetUserPath(D_HIRESTEXTURES_IDX);
+  const std::string cstex_bestdir = File::GetUserPath(D_HIRESTEXTURES_IDX) + game_id;
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: Same question, "best dir"?

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 6]
```diff
@@ -543,19 +543,36 @@ bool HiresTexture::LoadTexture(Level& level, const std::vector<u8>& buffer)
...
+  const std::string cstex_altdir = File::GetUserPath(D_HIRESTEXTURES_IDX) + cstex_altid;
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: Alt dir? Please find and pick a better name for this.

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 7]
```diff
@@ -543,19 +543,36 @@ bool HiresTexture::LoadTexture(Level& level, const std::vector<u8>& buffer)
...
+    INFO_LOG(VIDEO, "Custom textures directory '%s' does not exist, attempting region-free version",
+             game_id.c_str());
 
-  return texture_directory;
-}
+    if (File::Exists(cstex_altdir))
+    {
+      return cstex_altdir;
+    }
+    else
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: An `else` is unnecessary after a `return`.

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 8]
```diff
@@ -543,19 +543,36 @@ bool HiresTexture::LoadTexture(Level& level, const std::vector<u8>& buffer)
...
+    {
+      OSD::AddMessage(StringFromFormat("Missing: No custom textures for %s or %s were found!",
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: Please, no. I don't think people need to see an OSD message just because there are no custom textures. That's just useless noise.

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 9]
```diff
@@ -543,19 +543,36 @@ bool HiresTexture::LoadTexture(Level& level, const std::vector<u8>& buffer)
...
+                                       game_id.c_str(), cstex_altid.c_str()),
+                      20000);
+      WARN_LOG(VIDEO, "Missing: Custom textures directory '%s' does not exist.",
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: Doesn't warrant a warning log message either.

File: Source/Core/VideoCommon/HiresTextures.cpp [hunk 10]
```diff
@@ -543,19 +543,36 @@ bool HiresTexture::LoadTexture(Level& level, const std::vector<u8>& buffer)
...
+               cstex_altid.c_str());
+    }
+  }
 
-HiresTexture::~HiresTexture()
-{
+  return cstex_bestdir;
 }
 
+HiresTexture::~HiresTexture() {}
```
Comment (leoetlino) [inline Source/Core/VideoCommon/HiresTextures.cpp]: Unnecessary change.

File: Source/Core/VideoCommon/HiresTextures.cpp (hunk 1 above)
Comment (ZexaronS) [reply Source/Core/VideoCommon/HiresTextures.cpp]: Courtesy clang format 5.0.0

File: Source/Core/VideoCommon/HiresTextures.cpp (hunk 4 above)
Comment (ZexaronS) [reply Source/Core/VideoCommon/HiresTextures.cpp]: custom texture_alternative game identification

File: Source/Core/VideoCommon/HiresTextures.cpp (hunk 5 above)
Comment (ZexaronS) [reply Source/Core/VideoCommon/HiresTextures.cpp]: best case scenario

Comment (JosJuice) [reply Source/Core/VideoCommon/HiresTextures.cpp]: This scenario isn't really better than the other one...

**Comment by JosJuice** [comment]
This is not a good idea. Most people who use custom textures only have custom textures for some of their games rather than all of them, so this change would annoy them every time they start one of the games they don't have custom textures for, unless they go through the annoying process of enabling or disabling the custom texture setting for each game.

**Comment by ZexaronS** [comment]
@JosJuice Replaces one OSD message for another, nothing less, nothing more.

**Comment by JosJuice** [comment]
Then where's the part of this PR that makes the other OSD message not show up when there are 0 textures?

**Comment by leoetlino** [comment]
As far as I can tell, currently it just adds yet another message, instead of replacing the "loaded 0 textures" message.

**Comment by JosJuice** [comment]
Also, keep in mind that the "loaded 0 textures" message only shows up when prefetching is used. So even if you make that message not show up, the case where prefetching isn't used will still go from 0 OSD messages to 1 OSD message.

**Comment by ZexaronS** [comment]
Eventually ofcourse, I am talking about the state of when it gets merged, after I figure out or someone suggests something. Or is this not how "collaborative development" is suppose to work?

@JosJuice  Ah, ofcourse, silly me, I always use prefetching.


--------------
Contributing.md does not specify which clang-format version to use, I use 5.0.0 resulting in the above "unnecessary changes"

Do you want me to:

1. Revert the "unnecessary changes".
2. Switch to another clang format version.
3. Leave it like it is.

**Comment by JosJuice** [comment]
> Eventually ofcourse, I am talking about the state of when it gets merged, after I figure out or someone suggests something. Or is this not how "collaborative development" is suppose to work?

Sure, the PR doesn't have to be 100% done immediately. I just didn't understand that you meant that you were planning to do that later.

I still don't see the point of this change even if you fix that, though.

> Do you want me to:
>
> 1. Revert the "unnecessary changes".
> 2. Switch to another clang format version.
> 3. Leave it like it is.

Do it whichever way you'd like as long as it passes the lint bot. Options 1 and 2 should make it pass the lint bot, but 3 will definitely not.

**Comment by ZexaronS** [comment]
@JosJuice Allright https://buildbot.dolphin-emu.org/#/builders/22/builds/85 - it failed in #5996 too where I also used 5.0.0. 

Generally speaking:
What could potentially happen if lint was adjusted to how clang-format worked ? I suppose lint is the boss then, or at least right now.
What about the clang format config file, maybe this could be prevented there without having to revert to an old version of clang-format.

**Comment by Helios747** [comment]
You can run lint.sh which will tell you which version you need. Unrelated though, I wish it didn't require manual work on infra side to keep clang-format up to date.

Anyways, also mostly against this PR. This is all just useless noise you'd be implementing.

**Comment by stenzek** [comment]
Easy fix for clang-format version issues - apply the diff that the buildbot gives you back. It is slightly annoying due to the infra running an older version, but it's not like it takes a great deal of effort to work around.

-1 from me as well, I don't see the point in spamming the OSD when no textures are present. IMO, we should suppress the message entirely when no textures are found, because "0 textures loaded" is not very informative. And perhaps making the behavior similar between prefetching and no-prefetching modes. But that would be a change for further discussion.

**Comment by JosJuice** [comment]
> Edit: If this does go ahead, the OSD messages should be marked as translatable, AFAIK.

They shouldn't, because OSD messages won't be displayed if they contain non-ASCII characters.

**Comment by ZexaronS** [comment]
@stenzek I was planning on trying to get rid of the "0 textures loaded" yes.

@JosJuice Allright, what do we need to do to make that happen and solve it for good ?

**Comment by ZexaronS** [comment]
And with the latest changes:

1. Prefetcher and it's OSD messages won't run if there is no textures found.
2. OSD Message for no textures will trigger only if prefetching is used.
3. No texture WARN to INFO.

Translatable OSD Messages before this get's a go?

**Comment by stenzek** [comment]
> They shouldn't, because OSD messages won't be displayed if they contain non-ASCII characters.

Ah, good point. It might be a good idea to investigate this in the future then, perhaps after videocommon pipelines (moves the raster font rendering to common). Although pulling in a font rendering library might be kinda icky, the OSD messages IMO are probably good candidates for translation, since the user always sees them, like the rest of the UI. But feel free to argue against my logic here ;)

**Comment by stenzek** [comment]
I still don't agree with the idea of displaying an OSD message complaining about no custom textures being found. Many users (myself included) likely leave the option enabled in settings, while booting games without custom textures, thus it's just additional noise. But whatever, it's not a big enough deal for me to argue about.

**Comment by JosJuice** [comment]
> Ah, good point. It might be a good idea to investigate this in the future then, perhaps after videocommon pipelines (moves the raster font rendering to common). Although pulling in a font rendering library might be kinda icky, the OSD messages IMO are probably good candidates for translation, since the user always sees them, like the rest of the UI. But feel free to argue against my logic here ;)

I agree. Maybe we can add a `Host_` function that makes wx/Qt render the text to a buffer so that we don't need a separate library.

**Comment by ZexaronS** [comment]
@stenzek Right, why don't we view it as a notice then, rewrite it to feel more like that, not in a complaining way. Same way you get noticed about which renderer is used.

1. I fulfilled and put it from WARN to INFO level.
2. I fulfilled and made it so it's a replacement, it will show this instead of "0 textures loaded".
3. This happens earlier in boot so you're not being annoyed when you browse the menus.

That should make it less "annoying" than the existing messages, can anyone deny this now? 

I don't have a problem with disagreeing, but to the extent that it doesn't allow the PR to proceed. 

Another thing we didn't touch up on yet, do you all have prefetching enabled all the time too ?

**Comment by stenzek** [comment]
> I fulfilled and put it from WARN to INFO level.

I have no opinion on the logging level change. It's not relevant to the point I was trying to make, as the logging channels are disabled by default anyway, and won't be visible to users.

> If you used prefething you were already being annoyed by "0 textures loaded" anyway.

My point exactly. In this case, in my opinion, we should not display any message. Sure, go ahead and write it to the log. This makes sense, as the log is more verbose to begin with.

> Another thing we didn't touch up on yet, do you all have prefetching enabled all the time too ?

If I have custom textures enabled, yes.

**Comment by ZexaronS** [comment]
Update:

1. Changed INFO back to WARN
2. Removed OSD, now it won't show anything if no textures are found for whatever reason.

How about that?
